### ISSUE TITLE
hosts query taking a ton of RAM

### ISSUE THREAD

**Issue by therealmik** [kolide/fleet#2051]
On our system with just over 10k hosts, the hosts/all query takes around 6G of RAM, on a service that otherwise only uses 600M.

This seems to be related to it grabbing out all of the hosts into a `[]*kolide.Host` array, then serialising it into JSON by wrapping it into nested objects - this creates at-least two copies of the entire host array in memory, but seemingly more because it's using ~600k per host.

This is especially problematic because it's the first page loaded after login, and nobody really uses that page.

Ideas for fixing include:
* Default to "no hosts" on `hosts/manage`, and add an `all-hosts` slug for people that really want to load the whole list
* Stream results for the hosts endpoint (would require manually hacking together the JSON stream if you don't want to break compatibility)
* Make the front page something else (eg. packs)

I suspect #1980 is related to this issue.

**Comment by zwass** [comment]
The real fix to this, IMO, is to introduce proper pagination on the hosts page. There are a couple of other places where all of the hosts are loaded that are also unnecessary and should be addressed.

There has been support in the backend for pagination for quite some time, but it's a large effort to refactor the frontend to utilize this.

I do think it is worth looking into easy to implement alternatives for how we can make this page perform better. Perhaps limit the number of hosts loaded?

I do not think #1980 is related to this issue. In the case of loading query results, the full set of hosts is not loaded.

**Comment by therealmik** [comment]
An arbitrary limit seems reasonable - maybe just default to say 200?   If somebody wants to flick through more than 10 pages, we're probably doing them a favour :)

**Comment by zwass** [comment]
I dug into this and remembered something I had discovered previously: If we only load some of the hosts, we break other portions of the hosts page (for example, when a label is clicked not all of the hosts for that label will be loaded).

Ultimately, it's clear what needs to be done here but it's a pretty solid chunk of work and I hope that an organization with a large number of hosts (and hopefully some available budget) can sponsor that work.

**Comment by zwass** [comment]
I do have another stopgap measure: If there are over 200 hosts we don't reload the list of hosts from that page (which normally happens every 5 seconds). This reloading is kinda cool and useful when there is a small number of hosts so that it's easy to see new hosts coming in or changing online status, but is probably just a huge unnecessary resource consumption on both the browser and the server.

**Comment by zwass** [comment]
I implemented this stopgap in https://github.com/kolide/fleet/pull/2061.

**Comment by therealmik** [comment]
The stopgap certainly looks like it'd improve things.  I'll check back in once it's rolled into a release.

**Comment by zwass** [comment]
#261 is now merged and will be included in the next release.

**Comment by therealmik** [comment]
I upgraded to the new release, but it's still chewing a ton of RAM every time the hosts list is loaded.

Also, when I try to target a host for querying (in /queries/1234) the hostname I type appears briefly then gets replaced with some other thing and I get autocomplete suggestions that don't match my autocomplete at all.

**Comment by therealmik** [comment]
The target thing looks new and related to a different change - it's repeating the query with empty strings after performing the query for the preview text.

**Comment by HoustonDad** [comment]
I can confirm this is still occurring, and really causes a ton of pain points for anyone using Kolide with more than a few hundred hosts. With less than 700 hosts in Kolide, when loading the default dashboard / homepage (hosts/manage), it pulls a 52.0 MB hosts file. This file has 1,847,608 lines in it. This seems shocking, and perhaps an actual bug somehow?

One of the ideas @therealmik had was defaulting to "no hosts" on hosts/manage, and add an all-hosts slug for people that really want to load the whole list. This would be a great workaround until this can be fixed properly. 

I appreciate the work you did @zwass for the auto-reload, that's helped a lot. However for a lot of my team, who aren't on GigE connections like I am, this can cause a 30-60 second page load time, and I can't feel anyone could utilize that.

**Comment by therealmik** [comment]
I think if we removed the network interfaces from the all-hosts query, and instead put network interfaces on their own endpoint.

Then when rendering a host-box (or a set of host boxes) we could call the network interfaces query.

**Comment by zwass** [comment]
@JPerkster I explained above (https://github.com/kolide/fleet/issues/2051#issuecomment-500994816) why we can't disable loading all the hosts on the hosts page.

@therealmik I'd be open to reviewing a diff to achieve this, but I'm not sure that it's less effort than implementing the pagination on the frontend. If we did pursue this as a stopgap, I'd like to see it done in such a way that it doesn't make it more difficult to implement proper pagination in the future.

The other place that needs refactoring is the query page where the hosts need to be loaded to display in the target selection box

Ultimately most of the work I am doing on Fleet these days is on a volunteer basis and I can't work on features that exclusively benefit larger organizations without funding. I have a potential client for this one, but it takes time to work through the procurement process with larger organizations.
